eth/filters/bor_filter.go

```diff
@@ -0,0 +1,154 @@
+type BorBlockLogsFilter struct {
+	backend Backend
+
+	db        ethdb.Database
+	addresses []common.Address
+	topics    [][]common.Hash
+
+	block      common.Hash // Block hash if filtering a single block
+	begin, end int64       // Range interval if filtering multiple blocks
+}
+
+// NewBorBlockLogsRangeFilter creates a new filter which uses a bloom filter on blocks to
+// figure out whether a particular block is interesting or not.
+func NewBorBlockLogsRangeFilter(backend Backend, begin, end int64, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	// Create a generic filter and convert it into a range filter
+	filter := newBorBlockLogsFilter(backend, addresses, topics)
+	filter.begin = begin
+	filter.end = end
+
+	return filter
+}
+
+// NewBorBlockLogsFilter creates a new filter which directly inspects the contents of
+// a block to figure out whether it is interesting or not.
+func NewBorBlockLogsFilter(backend Backend, block common.Hash, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	// Create a generic filter and convert it into a block filter
+	filter := newBorBlockLogsFilter(backend, addresses, topics)
+	filter.block = block
+	return filter
+}
+
+// newBorBlockLogsFilter creates a generic filter that can either filter based on a block hash,
+// or based on range queries. The search criteria needs to be explicitly set.
+func newBorBlockLogsFilter(backend Backend, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	return &BorBlockLogsFilter{
+		backend:   backend,
+		addresses: addresses,
+		topics:    topics,
+		db:        backend.ChainDb(),
+	}
+}
+
+// Logs searches the blockchain for matching log entries, returning all from the
+// first block that contains matches, updating the start of the filter accordingly.
+func (f *BorBlockLogsFilter) Logs(ctx context.Context) ([]*types.Log, error) {
+	// If we're doing singleton block filtering, execute and return
+	if f.block != (common.Hash{}) {
+		receipt, err := f.backend.GetBorBlockReceipt(ctx, f.block)
+		if err != nil {
+			return nil, err
+		}
+
+		if receipt == nil {
+			return nil, errors.New("unknown block")
+		}
+		return f.borBlockLogs(ctx, receipt)
+	}
+
+	// Figure out the limits of the filter range
+	header, _ := f.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+	if header == nil {
+		return nil, nil
+	}
+	head := header.Number.Uint64()
+
+	if f.begin == -1 {
+		f.begin = int64(head)
+	}
+
+	// adjust begin for sprint
+	f.begin = nextSprintEnd(f.begin)
+
+	end := f.end
+	if f.end == -1 {
+		end = int64(head)
+	}
+
+	// Gather all indexed logs, and finish with non indexed ones
+	return f.unindexedLogs(ctx, uint64(end))
+}
+
+// unindexedLogs returns the logs matching the filter criteria based on raw block
+// iteration and bloom matching.
+func (f *BorBlockLogsFilter) unindexedLogs(ctx context.Context, end uint64) ([]*types.Log, error) {
```

Comment:
Not sure that mixing fields and passing argument is a good idea long term. You are passing end here as a part of the range, but you have .end field in the filters. At the same time inside .begin field is used. I would stick with using fields in this case.

Reply:
It's a common practice from normal log filter since `f.end` could be the latest block, and with subscription, `f.end` should remain the same throughout the subscription. That way same filter object can be used for future subscribed logs. I will look more into it and comment back.

I am making some changes in `eth_getLogs`, so this will be also used in the subscription. I will update you here.